Moved POCOs to EF and fixed tests

suhasj opened the pull request:
@HaoK I moved the POCOs to Identity.EF. The Identity.Test and Identity.InMemory.Test broke due to this change since they did not reference Identity.EF. Now we could either duplicate the code in the tests or reference the assembly and include the namespace. In addition the UserManagerTestBase is shared so that has to updated also. I choose to reference the assembly as this seems more cleaner. I made sure we reference only the POCOs and nothing else. Let me know if there is a better way of doing this.

/cc @divega 

You can skip the nitpick feedback for this PR

HaoK commented:
Nah, its a bad idea to add EF dependencies, just copy the pocos into the shared test code and rename them for the tests

suhasj commented:
In doing so, I have replicate the code in UserManagerTestBase for InMemory.Test project since it uses the IdentityUser object and that is a lot of duplication. Is that the right way ?

HaoK commented:
There should be no replication other than copying the pocos from Core to the Shared Test code along side user manager test base with a new name for the pocos

suhasj commented:
Ok let me see if the amount of duplicated code will be difficult to maintain in the future. I can reason out accordingly after that

suhasj commented:
Now if I duplicate the POCOs in the Shared folder and make UserManagerTestBase use them, the SqlStoreBase complains since it uses UserManagerTestBase with IdentityUser from Identity.EF. If I make all the test projects to use the POCOs in the shared folder, then we need to maintain them along with the POCOs in the source. My concern is that if we forget to update them then the tests will pass incorrectly.

HaoK commented:
Or maybe just move the IdentityUser classes to the new shared code stuff, so both EF and tests can depend on it safely.

HaoK commented:
Another route could be to look at what the tests are doing to access the pocos directly and instead replace them with calls to identity, which might let you remove the IdentityUser constraint entirely

suhasj commented:
Sure the second option might be an easy option. Let me try that out

suhasj commented:
@HaoK @divega 

Updated with changes
- UserManagerTestBase now is agnostic to user model and gets all attributes using manager apis. 
- Identity POCOs in Shared folder to be used by InMemory and Core tests
- updated samples

Question
- New extension method on ServiceCollection `AddIdentityWithTestPocos` which is used only in test because it uses the Shared Pocos while the one is source uses EF ones. Instead of this should i move the IdentityBuilderTests and IdentityOptionsTests to EF ? I would want to go with move option

divega commented on src/Microsoft.AspNet.Identity.EntityFramework/IdentityServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    public static class IdentityServiceCollectionExtensions
+    {
+        public static IdentityBuilder AddIdentity(this IServiceCollection services)
```
Interesting that these overload now moves here because the typed moved and hence they are tied to EF. I think it would probably better to get rid of them.

suhasj replied:
Fixed and updated

divega commented on src/Microsoft.AspNet.Identity.EntityFramework/IdentityServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+            return services.AddIdentity<IdentityUser, IdentityRole>();
+        }
+
+        public static IdentityBuilder AddIdentity(
```
Same as comment above.

suhasj commented:
Also one important feedback i need to getting username synchronously inside a Linq queries. For example 
`mgr.GetUserNameAsync(u).Result` in https://github.com/aspnet/Identity/pull/378/files#diff-bc862db146cf4a686c23d4fb0676c99eR584. The reason for doing this is that the Linq query cannot be async. Let me know if this is something that is not recommended. 

The workaround is to have a sync method on the `UserManagerTestBase` class which returns different user attributes. For example `GetUserId(user)`. This is abstract on the base and is implemented by the concrete test classes

divega commented on test/Shared/IdentityUser.cs:
```diff
@@ -87,5 +88,20 @@ public IdentityUser(string userName) : this()
+        /// <summary>
+        ///     Navigation property for users in the role
+        /// </summary>
+        public virtual ICollection<IdentityUserRole<TKey>> Roles { get; private set; } = new List<IdentityUserRole<TKey>>();
```
Will the `private` create issues with reflection in Mono?

HaoK replied:
Why do we need navigation properties on the shared test pocos?

suhasj replied:
True but I wanted to have parity with the Pocos in the source.

@divega for the private thing, the issue is fixed in the latest version of Mono. But should we revert it to protected ? Do we foresee users overriding this property considering it is a navigation property ?

HaoK replied:
Its not about parity, the EF pocos are EF specific, if the navigation properties are not needed for the tests, they shouldn't be there.

suhasj replied:
so the navigation properties are needed since the InMemory tests makes use of them. I need to keep them

divega commented:
> mgr.GetUserNameAsync(u).Result in https://github.com/aspnet/Identity/pull/378/files#diff-bc862db146cf4a686c23d4fb0676c99eR584. The reason for doing this is that the Linq query cannot be async. Let me know if this is something that is not recommended.

Can't you do this?
``` C#
var userName = await mgr.GetUserNameAsync(u);
Assert.Null(mgr.Users.FirstOrDefault(u => userName == "bogus"));
```

divega commented:
> New extension method on ServiceCollection AddIdentityWithTestPocos which is used only in test because it uses the Shared Pocos while the one is source uses EF ones. Instead of this should i move the IdentityBuilderTests and IdentityOptionsTests to EF ? I would want to go with move option

I don't understand this question, and I couldn't find the method right now. Did you already remove it? Were these test be independent of the provider?

HaoK commented on test/Microsoft.AspNet.Identity.EntityFramework.Test/DefaultPocoTest.cs:
```diff
@@ -65,6 +67,152 @@ public void DropDatabaseStart()
             IdentityResultAssert.IsSuccess(await userManager.DeleteAsync(user));
         }
 
+        [Fact]
```
Are these new tests or copied tests?

suhasj replied:
These are new ones I added to test if navigation properties work when used with `Include` on dbContext

HaoK commented on test/Microsoft.AspNet.Identity.InMemory.Test/InMemoryStoreTest.cs:
```diff
@@ -23,5 +23,31 @@ protected override void AddRoleStore(IServiceCollection services, object context
             services.AddSingleton<IRoleStore<IdentityRole>, InMemoryRoleStore<IdentityRole>>();
         }
+
+        protected override InMemoryUser CreateTestUser(string username = "", string email = "",
```
Isn't InMemoryUser the same as the new shared test user now?  Can we switch this over to use the common POCO?

suhasj replied:
Yeah we could do that. I'll InMemory user altogether. There is a TestUser also I believe. Can I remove that too ?

HaoK replied:
I would merge the new pocos with TestUser and keep the TestUser name as opposed to the confusing IdentityUser name

suhasj replied:
sure

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -83,15 +83,10 @@ protected RoleManager<TRole> CreateRoleManager(object context = null, IServiceCo
-            return new TUser() { UserName = namePrefix + Guid.NewGuid().ToString() };
-        }
+        protected abstract TUser CreateTestUser(string username = "", string email = "", string phoneNumber = "",
+            bool lockoutEnabled = false, DateTimeOffset? lockoutEnd = null);
```
I would go back to the namePrefix instead of allowing hardcoded user names, those almost certainly will cause issues accidentally if multiple tests pick the same user name

suhasj replied:
I kind of liked this that took the parameter as username and appended Guid when calling this. I feel strongly about keeping it this way since it more cleaner. 

Do you feel we are creating a pit of failure here for users ? They can always refer to test code for usage ?

HaoK replied:
This is test code not user code, you are creating a pit of failure for us, as if we write two tests that happen to choose the same user name now, they will fail/collide/ do bad things since the user name will exist.

suhasj replied:
There are some tests that need to remember the usernames for the next step, for example
```
var user = CreateUser(username)
result = await manager.FindByUserName(username)
```

For the above scenario I need to have the username at hand. Changing to namePrefix, it makes it tough to maintain the username. Earlier this was done using user.Username so it back to the same problem again. Can the overload take both username and flag that if true, uses what is passed ?

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -99,12 +94,12 @@ protected TRole CreateRole(string namePrefix = "")
 
-            Assert.Null(await manager.FindByIdAsync(user.Id.ToString()));
+            Assert.Null(await manager.FindByIdAsync(await manager.GetUserIdAsync(user)));
         }
```
nit: Get the user id once and store it

suhasj replied:
good point

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -139,18 +134,17 @@ protected TRole CreateRole(string namePrefix = "")
+            expectedLog = string.Format("{0} for user: {1} : {2}", "CheckPasswordAsync", await manager.GetUserIdAsync(user), false.ToString());
             IdentityResultAssert.VerifyLogMessage(manager.Logger, expectedLog);
-            Assert.True(await manager.CheckPasswordAsync(user, "New"));
+            Assert.True(await manager.CheckPasswordAsync(user, "Password1!"));
```
You accidentally broke this test, you deleted the HashPassword line above, password should get updated to New

suhasj replied:
No I replaced the `New` with `Password1!` since it doesn't meet the password requirement. Instead of directly setting the password hash I am doing it via `ChangePasswordAsync`

HaoK replied:
Read the test name more closely then :): "CanUpdatePasswordUsingHasher"

suhasj replied:
This is interesting, since the user object is now generic, we need to either have a SetPasswordHash on UserManager or abstract method on UserManagerTestBase to set it. I think I'll add it on the test base class. SetPasswordHash on UserManager seems not good

HaoK replied:
I think all of these methods showing up on the test base smells a bit, a better alternative might be to create a shared ITestUser internal interface and have both TestUser and IdentityUser implement this, then you can revert all of the new async manager calls and just use the old properties again since we can constrain against ITestUser now.

suhasj replied:
I can't have this constraint on the UserManagerTestBase generic since EF implementation IdentityUser would need to implement this interface as well. If we have the interface in test base, then in the EF tests, I need to have a TestUser which implements ITestUser and EF.IdentityUser but that doesn't seem that clean

HaoK replied:
What's wrong with having the real users implement an internal interface?

HaoK replied:
or even better, you can put the POCOs in internal shared code, and have the EF one extend and just make it public

internal class IdentityUserBase
public class EF.IdentityUser : IdentityUserBase { }

suhasj replied:
I like this solution. I'll have IdentityUserBase internal in Identity project all except the navigation properties. EF.IdentityUser extends this to add navigation properties and TestUser extends this to add properties as needed.

There is also an ApplicationUser in the test shared folder. Can I replace that too ?

divega replied:
If I am reading this correctly, it sounds very complicated, and it won't work :)

http://blogs.msdn.com/b/ericlippert/archive/2012/11/13/why-is-deriving-a-public-class-from-an-internal-class-illegal.aspx

I think it is actually fine to use the API on user manager as an abstraction layer and if we just need to do some trick to set the password hash alone, we can look into that more.


-------- Original message --------
From: Suhas Joshi <notifications@github.com>
Date: 03/03/2015 19:31 (GMT-08:00)
To: aspnet/Identity <Identity@noreply.github.com>
Cc: Diego Vega <Diego.Vega@microsoft.com>
Subject: Re: [Identity] [Feedback PR] Moved POCOs to EF and fixed tests (#378)


In test/Shared/UserManagerTestBase.cs<https://github.com/aspnet/Identity/pull/378#discussion_r25749952>:

>              IdentityResultAssert.VerifyLogMessage(manager.Logger, expectedLog);
> -            Assert.True(await manager.CheckPasswordAsync(user, "New"));
> +            Assert.True(await manager.CheckPasswordAsync(user, "Password1!"));


I like this solution. I'll have IdentityUserBase internal in Identity project all except the navigation properties. EF.IdentityUser extends this to add navigation properties and TestUser extends this to add properties as needed.

There is also an ApplicationUser in the test shared folder. Can I replace that too ?

-
Reply to this email directly or view it on GitHub<https://github.com/aspnet/Identity/pull/378/files#r25749952>.

divega replied:
>abstract method on UserManagerTestBase 

I like that more, personally. It means you can potentially derive from this class to test any provider without spoiling the entity types.

suhasj replied:
Going by the article we need to have IdentityUserBase public as well. Whether we do it just to unblock our tests, I am not sure. I feel we need a wider discussion to have a public user base class on Identity core. I'll go with the abstract method on UserManagerTestBase approach for now

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -506,27 +499,25 @@ protected TRole CreateRole(string namePrefix = "")
-            var user = CreateTestUser();
-            var user2 = new TUser() { UserName = user.UserName };
+            var user = CreateTestUser("AddDupeUserNameFails");
+            var user2 = CreateTestUser("AddDupeUserNameFails");
```
So if CreateTestUser only accepts name prefixes, just use SetuserName to explicitly set it to a fixed name after creating the user.

suhasj replied:
ok i'll change it if I revert to nameprefix :)

suhasj commented:
@divega Using a variable ! so obvious. To answer the second question, yes the class no longer exists to it is fine

suhasj commented:
@divega Now I recall that I did not use the variable since i need to use the object inside the `Where` clause. So `var userName = await mgr.GetUserNameAsync(u);` would not work since `u` is to be got from the lambda. The workaround I see is to use a foreach loop instead of `Where` which is not clean either.

divega commented:
Ah, I missed that. Sorry.

suhasj commented:
Updated with feedback
* Removed InMemoryUser and ApplicationUser. Identity.Test and IdentityInMemory.Test refer to TestUser only
* GetPasswordHash an abstract method on UserManagerTestBase
* CreateTestUser now takes namePrefix instead of username
* CreateTestRole does not take a namePrefix since most test need a expected rolename for further assertions.

suhasj commented:
@HaoK @divega 

Removing the feedback tag. I have addressed all feedback so wanted to make sure it is ready to check in

HaoK commented on src/Microsoft.AspNet.Identity.EntityFramework/IdentityUser.cs:
```diff
@@ -87,5 +88,20 @@ public IdentityUser(string userName) : this()
         ///     Used to record failures for the purposes of lockout
         /// </summary>
         public virtual int AccessFailedCount { get; set; }
+
```
You should double check with @divega but I'm pretty sure EF needs these nav properties to be public get/set to do their proxy stuff...?

suhasj replied:
All the properties now have public setters. I'll change these to public as well then

divega replied:
We actually don't have proxies in EF7 so far. @ajcvickers what do you think the best possible pattern would be for these collection navprops?

ajcvickers replied:
The way they are will work fine, but they don't need to be virtual (although you may want them to be virtual for other reasons) and they don't need to have a setter.

divega replied:
Thanks @ajcvickers. The setter on collection nav props actually looks very weird to me. @suhasj can we remove it?

HaoK commented on test/Shared/IdentityRoleClaim.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.AspNet.Identity.Test
+{
+    public class TestRoleClaim : TestRoleClaim<string> { }
```
Fix the file names for these shared files to match their classes

suhasj replied:
ah rebasing messed it up. I'll fix and verify the entire PR again

divega commented:
I'll take another look later today.

HaoK commented:
You might want to rebase since I added some functionals in InMemory yesterday, but otherwise looks :shipit: to me

suhasj commented:
@divega pinging a reminder

divega commented on test/Microsoft.AspNet.Identity.EntityFramework.Test/DefaultPocoTest.cs:
```diff
@@ -63,6 +65,152 @@ public void DropDatabaseStart()
+            });
+
+            var userManager = builder.ApplicationServices.GetRequiredService<UserManager<IdentityUser>>();
+            var dbContext = builder.ApplicationServices.GetRequiredService<IdentityDbContext>();
```
DbContext is IDisposable. Please use `using` even if doesn't seem needed right now.

divega replied:
Sorry, scratch that. I see now that you don't control the lifetime of this one.

divega replied:
However this still doesn't seem right. When will the context get disposed? I assume you copied this pattern from other tests?

suhasj replied:
I am assuming you are worried how the database is cleared ? There is a test `DropDatabaseStart` that is marked with lowest priority that will delete the database after all the tests in the class have run. This class existed all the way long. I have only added more tests that's all. Regarding the object, since it is created in the scope of the test, then it should be disposed at the end of it. If adding a using statement for DBContext is the recommended practice then sure I can update that but will have to do it in other tests as well

divega replied:
I am more worried about the risk of tests leaking resources, e.g. connections, or even memory depending how (well) the DI container implementation works and what we do in the tests.

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -592,8 +591,8 @@ protected TRole CreateRole(string namePrefix = "")
-            var usersQ = mgr.Users.Where(u => u.UserName.StartsWith("CanFindUsersViaUserQuerable"));
-            Assert.Null(mgr.Users.FirstOrDefault(u => u.UserName == "bogus"));
+            var usersQ = mgr.Users.Where(u => mgr.GetUserNameAsync(u).Result.StartsWith("CanFindUsersViaUserQuerable"));
+            Assert.Null(mgr.Users.FirstOrDefault(u => mgr.GetUserNameAsync(u).Result == "bogus"));
```
This one still seems weird. Using `Task<T>.Result` is pretty bad, and also this expression cannot be translated to a store query in EF so it will evaluate in memory. Isn't this actually an EF specific test that could depend on the EF POCOs? If we want to have it for any provider that implements IQueryable, let's chat about how to make it more general.

suhasj replied:
sure. The `.Result` was something i din't like too but ran out of ideas to make it better since the variable was inside the lambda. We can chat about it.

divega replied:
If we really think these test are general purpose and not EF only, e.g. if we think they have value for store providers that are not based on EF but support IQueryable, I propose that:

1. The tests for the IQueryable interfaces shouldn't fail if the IQueryable properties are not supported by the store, e.g. the test should test for the capability first.

2. The LINQ implementations are built to recognize simple patterns such as property accesses on the entities, and not calls to Identity. Since the test base class cannot depend on a specific set of entity types, we need to abstract this. We can do it easily by adding a set of virtual methods can invoke to build the expression that needs to be passed to the LINQ operators, e.g. for this case (brain-compiled):

``` C#
protected abstract Expression<Func<TUser, bool>> GetUserNamePredicate(string userName) 
{
}
``` 

An implementation for EF would be
``` C#
protected abstract Expression<Func<IdentityUser, bool>> GetUserNamePredicate(string userName) 
{
    return u => u.NormalizedUserName == userName;
}
```

suhasj replied:
This seems fair. I agree with the first point. For the second I want to minimize the number of abstract methods we have to fix this. I hope it would be only GetUserNamePredicate and GetRoleNamePredicate

suhasj replied:
For the first point, we should pass if the IQueryable is no implemented but rather skip it. I think it makes sense to have a custom xunit test attribute to skip the test. I'll prototype that but let me know if you guys think other wise

HaoK replied:
Eh?  Overkill, just check whether the store supports IQueryable and no-op pass the test if not...

HaoK replied:
I'm honestly not even sure we should build this functionality in at this point, given that we have no such stores :)

suhasj replied:
My thought was the user shouldn't get a false sense that the test is passing but in the skip attribute I needed access to the UserManager object which lead to generic hell. I think noop is the simplest option here

divega replied:
Since it seems we are going to keep the `IQueryable` properties it seems better to keep these tests here. I agree we should only have a few predicate building methods like this, but the pattern isn't bad.

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1096,34 +1091,36 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
-            Assert.Equal(roles.Count, manager.Roles.Count(r => r.Name.StartsWith("CanQuerableRolesTest")));
-            var r1 = manager.Roles.FirstOrDefault(r => r.Name.StartsWith("CanQuerableRolesTest1"));
+            Assert.Equal(roles.Count, manager.Roles.Count(r => manager.GetRoleNameAsync(r).Result.StartsWith("CanQuerableRolesTest")));
+            var r1 = manager.Roles.FirstOrDefault(r => manager.GetRoleNameAsync(r).Result.StartsWith("CanQuerableRolesTest1"));
```
Same comment as before about .Result. We should re-think how we test `IQueryables`.

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1208,7 +1205,7 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
                 foreach (var r in roles)
                 {
-                    Assert.True(rs.Any(role => role == r.Name));
+                    Assert.True(rs.Any(role => role == roleManager.GetRoleNameAsync(r).Result));
```
See other comments about `.Result`. Let's figure out if we can something better.

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1633,14 +1635,14 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
                 {
-                    IdentityResultAssert.IsSuccess(await manager.AddToRolesAsync(user, roles.Select(x => x.Name).AsEnumerable()));
-                    IdentityResultAssert.VerifyUserManagerSuccessLog(manager.Logger, "AddToRolesAsync", user.Id.ToString());
+                    IdentityResultAssert.IsSuccess(await manager.AddToRolesAsync(user, roles.Select(x => roleManager.GetRoleNameAsync(x).Result).AsEnumerable()));
```
Same about `.Result`.

suhasj commented:
Rebased with latest and remove all instances of `.Result` in the UserManagerTestBase. Added the predicate to compare user/roles names. Needed to add a contains flag since has two asserts that needed to test part of user/roles names. 

Also noop for IQueryable tests is the managers do not support them

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -82,15 +83,16 @@ protected RoleManager<TRole> CreateRoleManager(object context = null, IServiceCo
+
+        protected abstract TRole CreateTestRole(string roleName = "");
+
+        protected abstract Expression<Func<TUser, bool>> UserNamePredicate(string userName, bool contains = false);
```
1. I would rather simplify the tests to just succeed with simple equality and get rid of the contains flag on both predicates.
2. Contains is not the same as StartsWith.

divega replied:
Other than that, I think we are very close.

HaoK replied:
Yeah Contains != StartsWith, should be StartsWith

HaoK replied:
We can't simplify the tests to just do equality since we aren't cleaning up the database, so a lot of the tests use the same username prefix, like "test", and we only drop/recreate the database once per run.  Rather than trying to manually maintain user name uniqueness, I went with the approach of randomizing the names instead by always jamming on a guid, and then doing StratsWith instead of strict equality.

suhasj replied:
We were doing `name + guid` so I presumed Contains would still be valid check instead of StartWith considering we are looking if part of the string is present. I'll revert back to StartsWith it is just that I have to now maintain the original randomized username and pass it exactly to the predicate

divega replied:
I am ok with having StartsWith checks if we can't have just equality. But let's then split this in 4 simple predicate methods without any flags: UserNameEqualsPredicate, UserNameStartsWithPredicate, RoleNameEqualsPredicate, RoleNameStartsWithPredicate.

suhasj replied:
Added the four predicates and restored the test to original assert statements

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -505,43 +513,40 @@ protected TRole CreateRole(string namePrefix = "")
+            var user = CreateTestUser(email: "FooUser@yup.com");
+            var user2 = CreateTestUser(email: "FooUser@yup.com");
             IdentityResultAssert.IsSuccess(await manager.CreateAsync(user));
-            IdentityResultAssert.IsFailure(await manager.CreateAsync(user2), IdentityErrorDescriber.Default.DuplicateEmail(user.Email));
```
Why aren't we checking the failure code anymore?

suhasj replied:
thanks for catching. Copy paste error

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -568,17 +573,16 @@ protected TRole CreateRole(string namePrefix = "")
         [Fact]
         public async Task CanFindByEmail()
         {
+            var email = "foouser@test.com";
```
This is broken because you hardcoded an email

suhasj replied:
The requireuniqueemail flag is false so this is passing with all tests. We can have a random number with the email and set that flag to true

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1050,41 +1051,43 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
             var manager = CreateRoleManager();
-            var role = CreateRole("FindByIdAsync");
-            Assert.Null(await manager.FindByIdAsync(role.Id.ToString()));
+            var role = CreateTestRole("FindByIdAsync");
```
Another hardcoded name (this is why I think we should put the name randomization inside of the helper rather than expecting people to remember to do this...

suhasj replied:
we do add randomization in the helper when there is no argument provide. When something is, then we consider it to be fixed since they might need to compare that value later. I agree though that I missed it in this method

HaoK replied:
Also it doesn't make sense for CreateTestUser and CreateTestRole not to look the same, for the same reason CreateTestUser's was done the way it was, CreateTestRole should follow the same pattern

suhasj replied:
fixed. Both have the same pattern

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1093,39 +1096,45 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
             var manager = CreateRoleManager();
-            var roles = GenerateRoles("CanQuerableRolesTest", 4);
-            foreach (var r in roles)
+            if (!manager.SupportsQueryableRoles)
```
Isn't this backwards?  When would this test ever run today?

suhasj replied:
I fixed this for queryable users missed this for roles. Was doing some refactoring using Resharper but thanks for catching

HaoK commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1165,19 +1174,19 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
             var manager = CreateManager(context);
             var roleManager = CreateRoleManager(context);
-            var role = CreateRole("addUserTest");
+            var roleName = "AddUserTest" + Guid.NewGuid().ToString();
```
Yeah these role + guid name lines are all over, move this into the helper just like CreateTestUser

divega replied:
Sounds good to move this to the helper. If we need to reason about the generated value in a entity type-agnostic way, the helper can have output parameters for that.

suhasj replied:
Hao was ok with having a flag in the method that treats supplied prefix as the actual value else we append the guid. The default is to append the guid. There a few tests that need to do it and this approach is simpler than having an `out` parameter. You can look at the api and let me know your thoughts

divega replied:
Yeah sure, I guess we don't need to reason about the generated values in the test, which is fine.

suhasj commented:
Updated with 2 additional predicates to compare StartsWith for user/role name
* Some changes you might be interested in [predicates] (https://github.com/aspnet/Identity/pull/378/files#diff-bc862db146cf4a686c23d4fb0676c99eR94)
* queryable tests [1](https://github.com/aspnet/Identity/pull/378/files#diff-bc862db146cf4a686c23d4fb0676c99eR594) and [2] (https://github.com/aspnet/Identity/pull/378/files#diff-bc862db146cf4a686c23d4fb0676c99eR1097)

@ajcvickers made changes to EF for mapping. Should I rebase and send another PR to see if that works with these changes ?

suhasj commented:
I rebased it to latest EF which had renaming for mapping in OnModel creating. I have updated the IdentityDbContext with the latest changes

@ajcvickers can you please specifically review changes to [IdentityDbContext](https://github.com/aspnet/Identity/pull/378/files#diff-054beeacf6ff613faafca0aa8c1ad759R34)

divega commented on test/Shared/UserManagerTestBase.cs:
```diff
@@ -1093,39 +1097,44 @@ public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user
+                }
+                Assert.Equal(roles.Count, manager.Roles.Count(RoleNameStartsWithPredicate("CanQuerableRolesTest")));
+                var r1 = manager.Roles.FirstOrDefault(RoleNameStartsWithPredicate("CanQuerableRolesTest1"));
+                Assert.Equal(roles[1], r1);
```
Can we have a negativeassert as in the Users test and standardize on the same test data generation pattern?

suhasj replied:
sure I updated it and role name are done the same way as usernames.

ajcvickers commented:
IdentityDbContext changes look fine.

divega commented:
:shipit: to get it on beta4. We can address any outstanding test issues later.

suhasj commented:
thanks to everybody for patiently reviewing this
